### PR TITLE
feat(stepfunctions-tasks): Support for Athena APIs: StartQueryExecution, StopQueryExeuction, GetQueryResults and GetQueryExecution 

### PR DESCRIPTION
**Pull request by Sumeet-Badyal**
feat(stepfunctions-tasks): support for Athena APIs: StartQueryExecution, StopQueryExeuction, GetQueryResults and GetQueryExecution 

**Implementation**

Update package `@aws-cdk/aws-stepfunctions-tasks` to include support for Athena **StartQueryExecution**, **StopQueryExeuction**, **GetQueryResults**, **GetQueryExecution**    API as per documentation here: 
https://docs.aws.amazon.com/step-functions/latest/dg/connect-athena.html

Includes support for the following Amazon Athena API calls:
* `StartQueryExecution`
* `StopQueryExeuction`
* `GetQueryResults`
* `GetQueryExecution`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md [hunk 1]
```diff
@@ -205,6 +210,73 @@ The `EvaluateExpression` supports a `runtime` prop to specify the Lambda
...
+The [StartQueryExecution](https://docs.aws.amazon.com/athena/latest/APIReference/API_StartQueryExecution.html) API runs the SQL query statement.
+
+```ts
+import * as sfn from '@aws-cdk/aws-stepfunctions';
+import * as tasks from `@aws-cdk/aws-stepfunctions-tasks`;
+
+const startQueryExecutionJob = new tasks.AthenaStartQueryExecution(stack, 'Start Athena Query', {
+  queryString: sfn.JsonPath.stringAt('$.queryString'),
+  queryExecutionContext: {
+    catalog: ' ',
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/README.md:229]: this one might be worth a mini comment - it's not immediately clear why a user would want to do this.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+ * @experimental
+ */
+export interface EncryptionConfiguration {
+
+  /**
+   * Type of S3 server-side encryption enabled
+   *
+   * @default - 'SSE_S3'
+   */
+  readonly encryptionOption: string;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:228]: i believe in one of our other modules we call a similar property `serverSideEncryption` - does this need to be a string? could it be represented by an enum? I can't imagine that there are a lot of valid values here.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+  /**
+   * KMS key ARN or ID
+   *
+   * @default - No KMS key
+   */
+  readonly kmsKey?: kms.IKey;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:235]: these are typically named `<resource>EncryptionKey` or simply `encryptionKey` - can you rename to whichever one fits better for the usage?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+   * Type of S3 server-side encryption enabled
+   *
+   * @default - 'SSE_S3'
+   */
+  readonly encryptionOption: string;
+
+  /**
+   * KMS key ARN or ID
+   *
+   * @default - No KMS key
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:233]: is that right? doesn't it use the default master key if a customer provided key is not provided?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+   * S3 path of query results
+   *
+   * @default - No output location
+  */
+  readonly outputLocation?: string;
+
+  /**
+   * Encryption option used if enabled in S3
+   *
+   * @default - No encryption configuration
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:210]: describe the behaviour after the hyphen - i.e. SSE_S3 encrpytion is enabled with default encryption key.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+/**
+ * Location of query result along with S3 bucket configuration
+ *
+ * @see https://docs.aws.amazon.com/athena/latest/APIReference/API_ResultConfiguration.html
+ * @experimental
+ */
+export interface ResultConfiguration {
+
+  /**ResultConfigurationOptions
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:200]:
```suggestion
  /**
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+   * S3 path of query results
+   *
+   * @default - No output location
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:203]: where do query results go by default? are they just lost?
might also be handy to add `s3Location` in the name of the property. (there's a few examples of this in the repo)

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/get-query-execution.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+import { Construct } from 'constructs';
+import { integrationResourceArn, validatePatternSupported } from '../private/task-utils';
+
+/**
+ * Properties for getting a Query Execution
+ * @experimental
+ */
+export interface AthenaGetQueryExecutionProps extends sfn.TaskStateBaseProps {
+  /**
+   * Query that will be retrieved
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/get-query-execution.ts:12]: add an example query execution id if possible - it's helpful for customers to see `i.e. adfsaf-23trf23-f23rt23` to get a sense of what it represents in addition to the description provided in the doc string.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/get-query-execution.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+  constructor(scope: Construct, id: string, private readonly props: AthenaGetQueryExecutionProps) {
+    super(scope, id, props);
+    this.integrationPattern = props.integrationPattern ?? sfn.IntegrationPattern.REQUEST_RESPONSE;
+
+    validatePatternSupported(this.integrationPattern, AthenaGetQueryExecution.SUPPORTED_INTEGRATION_PATTERNS);
+
+    this.taskPolicies = [
+      new iam.PolicyStatement({
+        resources: ['*'], // Grant access to all workgroups as it can not be specified in the request @see https://docs.aws.amazon.com/athena/latest/ug/workgroups-iam-policy.html
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/get-query-execution.ts:43]: nit: the @see only needs to be there on public elements - simply the link is fine for inline comments

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+   * @default - No catalog
+   */
+  readonly catalog?: string;
+
+  /**
+   * Name of database used in query execution
+   *
+   * @default - No database
+   */
+  readonly database?: string;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:258]: should this property be named databaseName?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+ * @experimental
+ */
+export interface QueryExecutionContext {
+
+  /**
+   * Name of catalog used in query execution
+   *
+   * @default - No catalog
+   */
+  readonly catalog?: string;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:251]: should this property be named catalogName?

**Review by shivlaks — DISMISSED** [review]
@Sumeet-Badyal overall looks great! I have some minor suggestions and a couple of properties I think we should consider re-naming.

thanks for the high quality first cut!! I'm only about 2/3 of the way through this PR but figured I'd publish the comments so that you can iterate on them.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 7 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:203]: When one starts using Athena, one of the setup processes is under Settings to enter a Query Result Location for example: s3://query-results-bucket/folder/. This would be the default location but the user must create the bucket and key and can name it whatever they like.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 2 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:228]: In packages/@aws-cdk/aws-glue/lib/table.ts TableEncryption is exactly what I need. I am planning to import it.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 4 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:233]: If the default EncryptionOption `SSE_S3` is used I believe no KMS key is provided. If `SSE_KMS` or `CSE_KMS` then I believe you are correct the default master key is used. Would you like me to add something like "No KMS key for SSE_S3 and default master key for SSE_KMS and CSE_KMS"

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 7 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:203]: got it, maybe we can just describe that default instead

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 2 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:228]: I was mistaken this enum uses hyphen instead of an underscore. I'll make one and use it.

File: packages/@aws-cdk/aws-stepfunctions-tasks/README.md [hunk 12]
```diff
@@ -19,48 +19,55 @@ This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aw
...
-  - [Create Endpoint Config](#create-endpoint-config)
-  - [Create Model](#create-model)
-  - [Update Endpoint](#update-endpoint)
-- [SNS](#sns)
-- [Step Functions](#step-functions)
-  - [Start Execution](#start-execution)
-  - [Invoke Activity Worker](#invoke-activity)
-- [SQS](#sqs)
+- [Tasks for AWS Step Functions](#tasks-for-aws-step-functions)
+  - [Table Of Contents](#table-of-contents)
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/README.md:23]: what's the rationale for adding these? it's somewhat implied by the headers no?
the [CDK docs page](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-stepfunctions-tasks-readme.html#table-of-contents) renders based on headers so technically this TOC is more of a convenience.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+ *
+ * @see https://docs.aws.amazon.com/athena/latest/APIReference/API_EncryptionConfiguration.html
+ * @experimental
+ */
+export interface EncryptionConfiguration {
+
+  /**
+   * Type of S3 server-side encryption enabled
+   *
+   * @default - EncryptionOption.S3_MANAGED
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:227]:
```suggestion
   * @default EncryptionOption.S3_MANAGED
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/get-query-execution.ts (hunk 8 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/get-query-execution.ts:12]: sorry I was unclear... this was just an example - a real one (grabbed from your test looks like `2da557a1-7283-4c3d-8af9-058348f0bb02`)

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+        resources: [
+          cdk.Stack.of(this).formatArn({
+            service: 'athena',
+            resource: 'datacatalog',
+            resourceName: this.props.queryExecutionContext?.catalogName ? this.props.queryExecutionContext?.catalogName : 'AwsDataCatalog',
+          }),
+          cdk.Stack.of(this).formatArn({
+            service: 'athena',
+            resource: 'workgroup',
+            resourceName: this.props.workGroup ? this.props.workGroup : 'primary',
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:86]:
```suggestion
            resourceName: this.props.workGroup ?? 'primary',
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+  }
+
+  private createPolicyStatements(): iam.PolicyStatement[] {
+    const policyStatements = [
+      new iam.PolicyStatement({
+        resources: [
+          cdk.Stack.of(this).formatArn({
+            service: 'athena',
+            resource: 'datacatalog',
+            resourceName: this.props.queryExecutionContext?.catalogName ? this.props.queryExecutionContext?.catalogName : 'AwsDataCatalog',
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:81]:
```suggestion
            resourceName: this.props.queryExecutionContext?.catalogName ?? 'AwsDataCatalog',
```

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+            resourceName: '*', // access needed to query all tables https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
+          }),
+          cdk.Stack.of(this).formatArn({
+            service: 'glue',
+            resource: 'userdefinedfunction',
+            resourceName: '*', // users need access to start user defined functions from step functions https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
+          }),
+        ],
+      }),
+    );
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:158]: im unsure why permissions to `glue` api's are needed - the docs link leads to a list of topics but can you point to the specific ones that indicate that these policies are needed?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+   * @default - No catalog
+   */
+  readonly catalogName?: string;
+
+  /**
+   * Name of database used in query execution
+   *
+   * @default - No database
+   */
+  readonly databaseName?: string;
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:288]: curious: is a query execution context valid if neither is present?

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 18]
```diff
@@ -0,0 +1,289 @@
...
+        actions: ['glue:BatchCreatePartition',
+          'glue:BatchDeletePartition',
+          'glue:BatchDeleteTable',
+          'glue:BatchGetPartition',
+          'glue:CreateDatabase',
+          'glue:CreatePartition',
+          'glue:CreateTable',
+          'glue:DeleteDatabase',
+          'glue:DeletePartition',
+          'glue:DeleteTable',
```
Comment (shivlaks) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:126]: does this mean that any glue database/table can be deleted? 
the permissions here seem really wide

**Review by shivlaks — DISMISSED** [review]

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 17 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:288]: I am going to provide two use cases and then hopefully that will clarify things:
relevant link https://docs.aws.amazon.com/IAM/latest/UserGuide/list_awsglue.html

1. Start an an Athena Query that creates a table, this requires write permissions to catalog, database and table as shown in the actions table in the link above.
2. Start an Athena Query that gets a userdefinedfunction(UDF) then runs the function. This requires read permissions to a UDF

In the table Resource types defined by AWS Glue the table shows how to scope down permissions if one has the UDF or table name which I don’t.

Basically any athena query possible must be able to run from a state machine hence the large permissions granted. 

One thing I can do is that in the table Resource types defined by AWS Glue, the permissions are scoped down if the user sends an optional parameter databaseName.
i.e. currently: UDF /glue/user-defined-function/*
changed to: /glue/user-defined-function/DATABASENAME/*
or if no database name provided /glue/user-defined-function/default/*

This will prevent cross database access.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 18 above)
Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:126]: Refer to https://github.com/aws/aws-cdk/pull/11045#discussion_r512174712

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts (hunk 17 above)
Comment (shivlaks) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:288]: got it, thanks for clarifying. I think we should seek to simplify this. it's unfortunate we don't have L2 constructs for Athena yet, as `grant*` APIs would/should be leveraged here. 

let's introduce any changes that will ensure least privilege for now.

Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:288]: I believe those changes have been included in the commit found at these two links:

https://github.com/aws/aws-cdk/pull/11045/files#diff-edd80b0b061b6b337000af6848daeda4011822de72da5749dbcd430c8089bc37R149
https://github.com/aws/aws-cdk/pull/11045/files#diff-edd80b0b061b6b337000af6848daeda4011822de72da5749dbcd430c8089bc37R154

Granting access to all tables and user defined functions in the specified database or the default database if no database has been included in the request.

**Review by shivlaks — APPROVED** [review]
nice work @Sumeet-Badyal !!

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts [hunk 19]
```diff
@@ -0,0 +1,289 @@
...
+      new iam.PolicyStatement({
+        actions: ['s3:AbortMultipartUpload',
+          's3:CreateBucket',
+          's3:GetBucketLocation',
+          's3:GetObject',
+          's3:ListBucket',
+          's3:ListBucketMultipartUploads',
+          's3:ListMultipartUploadParts',
+          's3:PutObject'],
+        resources: [this.props.resultConfiguration?.outputLocation ?? '*'], // Need S3 location where data is stored https://docs.aws.amazon.com/athena/latest/ug/security-iam-athena.html
```
Comment (akiym) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:104]: Since these policy resources expect ARN, passing `outputLocation` (S3 URL like `s3://...`) will get an error.
Even if `outputLocation` is specified to props, we may have to pass `*` to resources or parse S3 URL to get ARN correctly.

Comment (Sumeet-Badyal) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/athena/start-query-execution.ts:104]: Thank you for spotting this. I parsed the s3 bucket correctly like you suggested but still got an error `Unable to verify/create output bucket`. When I reran the exact same state machine but with wildcard permissions the machine succeeded and sent the logs to the correct output location. I will be submitting a fix immediately granting wildcard permissions for S3 and LakeFormation
